Make the package installable.

randrej opened the pull request:
Don't merge this until it's thoroughly tested. I tested it superficially.

I did the following:
- moved Python sources to package folder called `kintree` as per the packaging convention
- moved the configuration and cache (search, image) files to folders provided by `appdirs`
- added all dependencies to `pyproject.toml`, which means we may now build or install the package via Poetry
- removed all submodules and replaced them with dependencies in `pyproject.toml`
- added `kintree` and `kintree_setup_inventree` entrypoints to the package: pip installing the package adds these commands to the PATH

I haven't changed the readme yet, nor added the PyPI deployment stuff. Didn't run or touch the tests.

You need to install Poetry now to do development / building. Install poetry outside any venv. Poetry manages the venv, the requirements, package building.  I'm not using a classic `setup.py`.

## Development
To make the local virtualenv for development, run `poetry install --no-root`. To enable it, type `poetry shell`. Alternatively, you may run commands in the venv with `poetry run <command>" without entering the venv.

Since I made the imports relative in all source files, running the package with `python kintree/kintree_gui.py` no longer works. To run the program in the venv you have to type `python -m kintree.kintree_gui`.

## Building the package
To build the package, just run `poetry build`. The resulting package is created in the `dist` subfolder, and may be installed system-wide with `pip install dist/<something>`.  On Linux, you should use `pip install --user`. Once the package is installed, you can run `kintree` on the command line, anywhere (granted that Python path is properly installed).

The version is specified in `pyproject.toml`. I can probably later get poetry to ignore this version and use the latest Git tag as the version, but I'm not sure how well this works with poetry since it's a hack; it worked OK w/ `setup.py`.

## PyPI deployment
I'll add code to push this to pip in CI later, but we'll need some realtime coordination to do this, since you have to make a PyPI account, generate secrets for this package and save them to Github Actions. Gitter / IRC / Signal? I don't have social media accounts apart from LinkedIn. Afterwards, just pushing the tag will trigger testing an Python deployment.

eeintech commented:
@randrej Ok I finally was able to give it a shot. I followed your instructions and was able to run `kintree` installing from the wheel, great stuff!

I have few questions:
- `appdirs` seems appealing but I don't see a good way to define the `user_config_dir` and `user_cache_dir` manually. I'm actually thinking to contain those files within one folder set by the user within the Ki-nTree application, then it would be easier to retrieve those files. Is that feasible?

> removed all submodules and replaced them with dependencies in `pyproject.toml`
- I see the submodule versions are preceded by a `^`, that does it mean? Are those submodule loaded from their Git repo directly?

I also had to force installing the `wrapt_timeout_decorator` package directly from GitHub:
https://github.com/sparkmicro/Ki-nTree/blob/8cd8780f95abb4cbe9b2135141ba154f177a5508/tasks.py#L16

Which I did in a different way than installing as submodule. Hoping the `^` has the same effect!

- I noticed that you have skipped the `invoke` package, is it because `Poetry` replaces it? Can I run custom commands with `Poetry` too?

randrej commented:
> appdirs seems appealing but I don't see a good way to define the user_config_dir and user_cache_dir manually. I'm actually thinking to contain those files within one folder set by the user within the Ki-nTree application, then it would be easier to retrieve those files. Is that feasible?

You don't define them manually, the URLs are retrieved by the `appddirs` module. If you think these folders are hard to find manually, we can add the calculated values somewhere in the GUI (somewhere in settings?), so that the users know where exactly these are. Otherwise, we can write the URLs in the readme:

#### Linux
*configuration*: `$XDG_CONFIG_HOME/kintree`, which is almost always `~/.config/kintree`  
*cache*: `$XDG_CACHE_HOME/kintree`, which is almost always `~/.cache/kintree`

#### Windows
*cache* and *configuration*: `%LocalAppData%\sparkmicro\kintree`, which defaults to `C:\Users\<username>\AppData\Local\sparkmicro\kintree` though you can just paste the first URL in Explorer and it resolves.

---

>>   removed all submodules and replaced them with dependencies in pyproject.toml
>
>   I see the submodule versions are preceded by a ^, what does it mean? Are those submodule loaded from their Git repo directly?

The projects that were included via submodules all have PyPI distributions, so I'm using their latest PyPI versions. The `^` is a [caret requirement](https://python-poetry.org/docs/dependency-specification/#caret-requirements), a fancy way of version range specification in Poetry (and in general). It works really well for projects versioned with [semver](https://semver.org/), which is most projects nowadays, because the last digit group signifies non-breaking changes.

---

> I also had to force installing the wrapt_timeout_decorator package directly from GitHub
> Which I did in a different way than installing as submodule. Hoping the ^ has the same effect!

I'm not installing it from Github, just the latest version in PyPI. Poetry also allows directly installing from Github if that is really necessary, but I think it also works with this version. Is there any reason to prefer the master branch on Github to this version on PyPI?

---

> I noticed that you have skipped the invoke package, is it because Poetry replaces it? Can I run custom commands with Poetry too?

No, Poetry doesn't provide such funcitonality, Invoke shouldn't be replaced for running commands. However, since all of the commands are install-related, and I've moved all the install steps to Poetry, I'm not sure it's necessary at the moment - all those commands are no longer needed.

eeintech commented:
> If you think these folders are hard to find manually, we can add the calculated values somewhere in the GUI (somewhere in settings?)

That could work. And yes I think those are hard to find and given I did not implement all the settings in the UI, you still need to access those files quite regularly... Long-term nobody should ever need to manually edit them but it's a long way before that happens (at least it is not my primary focus and don't have the time to refactor the UI much).

> The `^` is a [caret requirement](https://python-poetry.org/docs/dependency-specification/#caret-requirements), a fancy way of version range specification in Poetry (and in general).

Very cool, I like that caret! I always found it tedious to have to write something like `>=1.x.x,<2.0.0` for all the dependencies.

> Is there any reason to prefer the master branch on Github to this version on PyPI?

There was something about the latest pip release that was not working quite right with that timeout library, so had to pull master on Github instead which had the fix for it. But can't remember exactly what 🤔  Would have to reproduce it...
Good to know we can use Github though.

> However, since all of the commands are install-related, and I've moved all the install steps to Poetry, I'm not sure it's necessary at the moment - all those commands are no longer needed.

That's right, also for testing too. But those could go away with a better CI implementation... 

@randrej If you have time to contribute to Ki-nTree, I think the best would be to work towards a `0.4.0` release with those goals in mind:

* Make it available as a `pip` package (#32 and #35) using `Poetry`
* Show `cache` and `settings` directories in the UI settings (AppDir)
* Setup code formatting rules and checks (#33)
* Move CI to Github actions and improve testing (https://github.com/sparkmicro/Ki-nTree/issues/32#issuecomment-821986128)
* Fix self-certificate issue (#36)
* Make sure all feature of latest 0.3.x version work (I will probably release 0.3.7 this week but I don't foresee any immediate changes after that)

I would keep the UI migration (#37) for later, although I'm quite excited about it 😃
What do you think?

randrej commented:
Hi, sorry for replying with such a long delay, I'm quite busy.
I want to contribute and I agree with your proposed plan for 0.4.0. The issue is that I don't have very much free time at this point, but I will in the next couple of weeks - certainly not this week. If this is fine for you, then let's do it.

eeintech commented:
@randrej Awesome! The timing is no problem, I'm also quite busy those days and this is more like a side project 😃

eeintech commented:
Items 3 and 4 are now done :smiley:

eeintech commented:
@randrej For the user config files and cache directories, I actually thought to go my own way in #41 instead of the AppDir package, it does not seem that well maintained and I still wanted users to be able to tell where their configuration is stored. It creates a `user_config.yaml` files at the root of the project, which should be fine when it is packaged I believe too. This file maps the path for user files and cache. It can be changed through the UI, as described in #41

randrej commented:
@eeintech Sure, this way is completely fine. AppDir hasn't been updated recently, but it's feature-complete enough. The way you wrote it also works, but I suggest using ~/.config/kintree on Linux. Here's my suggestion in code:

https://github.com/sparkmicro/Ki-nTree/pull/41#pullrequestreview-699365729

randrej commented:
I'll add a PR for the self-cert issue, and pack everything up with Poetry.

eeintech commented:
I have merged #42 which fixes the comment/review you sent me :smiley: 

Now I'll let you do the certificate fix and test it cause I don't really know the deal.

If you can package that's great too!

randrej commented:
I've recreated the pull request on top of the new master branch, and removed obsolete changes. I have to fix the tests, but this should complete the majority of 0.4.0 requirements.

I suggest we do the following:

1. I finish the tests, write a temporary readme for installing with poetry, make the package buildable and installable by Poetry. I can probably do it in a day.
2. You merge these changes. This is important to me, because the folder structure of this branch is different and there are relative imports now - if you add any changes I have to merge into my branch, it's a lot of work for me.
3. I proceed to make Github actions for pushing to TestPyPI - a fake PyPI instance for testing deployment. Once this is done, you create credentials for the real PyPI and we set the package to be deployed on the real PyPI automatically. This requires some cooperation between the two of us - I suggest something realtime, like IRC or Gitter.
4. I make the Arch Linux package for the program.

rcludwick commented on pyproject.toml:
```diff
@@ -0,0 +1,50 @@
+certifi = ">=2020.6.20,<2021.0"
+cffi = ">=1.14.1,<2.0"
+chardet = ">=3.0.4,<4.0"
+coverage = ">=5.2.1,<6.0"
```
put this in the dev dependency.

randrej replied:
@eeintech Please review this and check which of these packages are actually used, either for dev / test purposes, or for actual implementation.

For example, I can't seem to find any use of the following packages:

- certifi
- cffi
- chardet
- coveralls
- inflection
- pycparser

eeintech replied:
@randrej Apparently `certifi`, `cffi`, `chardet` and `inflection` are all dependencies of the Digi-Key API project. But it seems like `pycparser` serves no purpose so I removed it :+1:

eeintech replied:
`coverage` can be installed when the tests are run only using the `inv test` command

randrej replied:
> @randrej Apparently `certifi`, `cffi`, `chardet` and `inflection` are all dependencies of the Digi-Key API project. But it seems like `pycparser` serves no purpose so I removed it +1

If that's the case, they can be removed from this package's dependency list. They will be installed as transitive dependencies anyway. How about the other ones? Are there any other packages that are dependencies of packages formerly included as submodules?

eeintech replied:
Ok I did a big clean-up, you are right, no need of maintaining dependencies from other packages!

rcludwick commented on pyproject.toml:
```diff
@@ -0,0 +1,50 @@
+cffi = ">=1.14.1,<2.0"
+chardet = ">=3.0.4,<4.0"
+coverage = ">=5.2.1,<6.0"
+coveralls = ">=2.1.2,<3.0"
```
This one too.

randrej replied:
@eeintech Look at this, too.

rcludwick commented on pyproject.toml:
```diff
@@ -0,0 +1,50 @@
+inflection = ">=0.5.0,<1.0"
+inventree = ">=0.2.4,<1.0"
+pycparser = ">=2.20,<3.0"
+pyinstaller = ">=4.0,<5.0"
```
Technically pyinstaller-* are dev dependencies.

randrej replied:
@eeintech This, too.

eeintech replied:
So do you mean removing this dependency and installing it only when the `inv build` is ran? I would like to keep the `inv build` available for all users.

eeintech commented:
@randrej Ok sounds reasonable strategy to me! Let me know when you are ready to merge and we'll push forward with TestPyPI

randrej commented:
Please run all the tests yourself, see if everything works, and check if the readme is clear enough. Also please check which packages can be removed from the dependency list. Let's remove all the unneeded requirements before merging.

Once this is merged, we can start preparing for PyPI publishing.

randrej commented:
@eeintech I looked into publishing the package to PyPI, and PyPI only allows requirements on PyPI itself - git requirements won't work. This means we have 3 options when it comes to [wrapt_timeout_decorator](https://github.com/bitranox/wrapt_timeout_decorator):

1. Use the PyPI version, fix any issues locally by altering the way we call it or even monkeypatching the code.
2. Ask the person developing it to publish a new version to PyPI
3. Add it as a submodule here, ship the code inside `kintree`.

What exactly doesn't work with the latest PyPI version? What's the exact issue?

eeintech commented:
@randrej Ok I'll test this out as soon as I can, thanks for all the good work!

For the timeout package, I will take another look. I know there was an issue with the PyPI release but it's not fresh in my mind what it was, it's possible we can fix it locally. And there are still missing to push their latest code to PyPI...

rcludwick commented:
It's MIT license, so I would just bring it into this project, with an
intent on replacing it later.

On Mon, Jul 12, 2021, 5:56 AM Francois ***@***.***> wrote:

> @randrej <https://github.com/randrej> Ok I'll test this out as soon as I
> can, thanks for all the good work!
>
> For the timeout package, I will take another look. I know there was an
> issue with the PyPI release but it's not fresh in my mind what it was, it's
> possible we can fix it locally. And there are still missing to push their
> latest code to PyPI...
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/sparkmicro/Ki-nTree/pull/35#issuecomment-878216942>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAL67I2PBM65QFVLNNFYTPDTXLJ7PANCNFSM43DEMLYA>
> .
>

eeintech commented:
@randrej Unfortunately there is no way to share my repo secrets to your fork and therefore the Digi-Key API test can't be ran in Actions. The good news is that I have tested this locally and it works great!

I went through the README, I think it's good enough for now but for sure the installation feels like more complicated at this point!
Can't wait to see the `pip install kintree` as the only instruction :smile:

eeintech commented:
@randrej Resolved all the review comments from @rcludwick so now the list of dependencies is trimmed down to the minimum (I think).

I added the latest `wrapt_timeout_decorator` code in `kintree/wrapt_timeout_decorator` folder and updated the methods using it.

Any other thing missing before I merge it?

randrej reviewed (commented):
This all looks OK.

randrej commented:
> @randrej Unfortunately there is no way to share my repo secrets to your fork and therefore the Digi-Key API test can't be ran in Actions. The good news is that I have tested this locally and it works great!
> 
> I went through the README, I think it's good enough for now but for sure the installation feels like more complicated at this point!
> Can't wait to see the `pip install kintree` as the only instruction smile

I added my own secrets to my fork (copied from my own Ki-nTree config), and for some reason they don't work (they appear to be missing).




> @randrej Resolved all the review comments from @rcludwick so now the list of dependencies is trimmed down to the minimum (I think).
> 
> I added the latest `wrapt_timeout_decorator` code in `kintree/wrapt_timeout_decorator` folder and updated the methods using it.
> 
> Any other thing missing before I merge it?

I've reviewed the code and everything seems to be alright. I don't see anything missing for this release.

The next release should be on PyPI. I'll make a new issue for finally setting up an action for pushing to PyPI.

eeintech commented:
I re-ran the setup and noticed the version discrepancy with the latest from master, it's corrected now :smiley: 

Merging!
